### PR TITLE
Psalm it up!

### PR DESCRIPTION
**Pull request by djmetzle**
Let's use Psalm to Type check and annotate our new library.

Psalm immediately found some bugs, and a bunch of missing types.

The most important thing for our upstream consumption of the library is template types, ie. `Option<T>`. This will let us be much safer.

https://github.com/iFixit/php7-optional/blob/658312e8d556d5ec715f3e2071d4b20524c20a53/src/Option.php#L7-L19

Also, let's add Psalm to CI so that we'll catch any future errors.

I've left annotating away all the INFO warnings as an exercise for the library maintainer.

CC @jarstelfox

### CODE REVIEW

File: src/Option.php [hunk 1]
```diff
@@ -4,36 +4,45 @@
...
+   /**
+    * @param mixed $alternative
+    * @return mixed
+    **/
    public function valueOr($alternative) {
       return $this->hasValue ? $this->value : $alternative;
    }
 
+   /**
+    * @return mixed
```
Comment (djmetzle) [inline src/Option.php]:
```suggestion
    * @return T
```

File: src/Option.php [hunk 2]
```diff
@@ -4,36 +4,45 @@
...
+    **/
    public function valueOrCreate(callable $func) {
-      $this->throwIfNull($func);
-
       return $this->hasValue ? $this->value : $func();
    }
 
+   /** @param mixed $alternative */
```
Comment (djmetzle) [inline src/Option.php]:
```suggestion
   /** @param T $alternative */
```

File: src/Option.php [hunk 3]
```diff
@@ -4,36 +4,45 @@
...
    }
 
-   public function hasValue() {
+   public function hasValue(): bool {
       return $this->hasValue;
    }
 
+   /**
+    * @param mixed $alternative
+    * @return mixed
```
Comment (djmetzle) [inline src/Option.php]:
```suggestion
    * @return T
```

File: src/Option.php [hunk 4]
```diff
@@ -4,36 +4,45 @@
...
       $this->value = $value;
    }
 
-   public function hasValue() {
+   public function hasValue(): bool {
       return $this->hasValue;
    }
 
+   /**
+    * @param mixed $alternative
```
Comment (djmetzle) [inline src/Option.php]:
```suggestion
    * @param T $alternative
```

**Comment by jarstelfox** [comment]
CR 🌵 still more to teach psalm, but thank you for adding this!

**Pull request merged**